### PR TITLE
bug: Fix high memory consumption when using Junit formatter

### PR DESCRIPTION
**Pull request by simon-auch**
The high memory consumption is caused by the formatter keeping many events for the lifetime of a feature.
Some events contain references to the BehatContext, resulting in the BehatContext not being dropped.
If a feature contains many scenarios/examples and the BehatContext is big, this can lead to OoM errors.

This commit drastically reduces the amount of events the Junit formatter keeps and limits the lifetime of those it keeps to the lifetime of an example.

This PR fixes https://github.com/Behat/Behat/issues/1103

### CODE REVIEW

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php [hunk 1]
```diff
@@ -178,16 +135,29 @@ public function printFeatureOnAfterEvent(Formatter $formatter, Event $event)
...
 
-        $this->featurePrinter->printFooter($formatter, $event->getTestResult());
-        $this->afterScenarioTestedEvents = array();
+    /**
+     * Prints the feature on AFTER event.
+     *
+     * @param Formatter $formatter
+     * @param Event     $event
+     */
+    private function printFeatureOnAfterEvent(Formatter $formatter, Event $event)
```
Comment (simon-auch) [inline src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:157]: This method was public before, but not actually part of the interface.
Should this be kept public?

**Comment by simon-auch** [comment]
This PR is still missing a feature (as per the contribution guidelines), I'm not sure however what the best way to test this would be.

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php [hunk 2]
```diff
@@ -154,21 +117,15 @@ private function captureStepEvent(Event $event)
...
+     * Captures scenario tested event.
      *
      * @param Formatter $formatter
-     * @param Event     $event
+     * @param ScenarioTested $event
      */
-    public function printFeatureOnAfterEvent(Formatter $formatter, Event $event)
+    private function printScenarioEvent(Formatter $formatter, Event $event)
     {
-        if (!$event instanceof AfterFeatureTested) {
```
Comment (stof) [inline src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:164]: I'd rather keep such early return

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php [hunk 3]
```diff
@@ -154,21 +117,15 @@ private function captureStepEvent(Event $event)
     }
 
     /**
-     * Prints the feature on AFTER event.
+     * Captures scenario tested event.
      *
      * @param Formatter $formatter
-     * @param Event     $event
+     * @param ScenarioTested $event
```
Comment (stof) [inline src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:123]: adding a phpdoc that does not match the parameter type does not make sense

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php (hunk 2 above)
Comment (simon-auch) [reply src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:164]: reverted to use early returns again.

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php (hunk 3 above)
Comment (simon-auch) [reply src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php]: fixed

File: src/Behat/Testwork/Output/Printer/JUnitOutputPrinter.php [hunk 4]
```diff
@@ -85,6 +85,15 @@ public function addTestsuite(array $testsuiteAttributes = array())
         $this->addAttributesToNode($this->currentTestsuite, $testsuiteAttributes);
     }
 
+    /**
+     * Extends the current <testsuite> node.
+     *
+     * @param array $testsuiteAttributes
+     */
+    public function extendTestsuiteAttributes(array $testsuiteAttributes = array()){
```
Comment (dkarlovi) [inline src/Behat/Testwork/Output/Printer/JUnitOutputPrinter.php:93]: Code style, `{` on newline.

Comment (simon-auch) [reply src/Behat/Testwork/Output/Printer/JUnitOutputPrinter.php]: fixed

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php (hunk 1 above)
Comment (ciaranmcnulty) [reply src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:159]: I think that's fine if it's not part of the interface, if anyone complains we can revert

**Comment by ciaranmcnulty** [comment]
Thanks @simon-auch

**Comment by dkarlovi** [comment]
@stof this seems like it could be merged.

**Review by dkarlovi — APPROVED** [review]

**Comment by jorandirkzwager** [comment]
Will these changes be merged someday?

**Comment by simon-auch** [comment]
Is there anything I can do to help get these changes merged?

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php [hunk 5]
```diff
@@ -154,40 +118,49 @@ private function captureStepEvent(Event $event)
     }
 
     /**
-     * Prints the feature on AFTER event.
+     * Captures scenario tested event.
```
Comment (stof) [inline src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:121]: the method prints something. It does not capture something

File: src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php [hunk 6]
```diff
@@ -154,40 +118,49 @@ private function captureStepEvent(Event $event)
...
+        $afterScenarioTested = $event;
+        $this->scenarioPrinter->printOpenTag(
+            $formatter,
+            $afterScenarioTested->getFeature(),
+            $afterScenarioTested->getScenario(),
+            $afterScenarioTested->getTestResult(),
+            $event->getFeature()->getFile()
+        );
+
+        /** @var AfterStepSetup $afterStepSetup */
```
Comment (stof) [inline src/Behat/Behat/Output/Node/EventListener/JUnit/JUnitFeatureElementListener.php:140]: This should be useless thanks to the phpdoc on the property describing the type of items in the array (which is currently described as `AfterSetup`, not `AfterStepSetup`, so one of the 2 places has an issue IMO.

**Comment by thomandre** [comment]
How can we help solving this issue?

**Comment by carlos-granados** [comment]
@simon-auch many thanks for your contribution. I have merge the latest master into the branch and all tests are passing so this is almost ready to go, just a couple of things needed:
- Could you please incorporate the latest change suggested by @stof ?
- Could you please remove the change to the CHANGELOG? We are now handling this separately

**Review by carlos-granados — CHANGES_REQUESTED** [review]
@simon-auch see my comment yesterday

**Review by carlos-granados — APPROVED** [review]

**Comment by carlos-granados** [comment]
@acoulton since we did not get any response from the original author I made the two remaining changes (removing the CHANGELOG changes and adding a more detailed type hint) myself. I think this should be ready to go, could you take a new look? Thanks!

**Review by acoulton — APPROVED** [review]
LGTM, thanks @carlos-granados and @simon-auch

**Pull request merged**